File: 백준/Silver/1149. RGB거리/RGB거리.py
# dp[i][j]
# dp[i][0] 빨   dp[i][1] 초    dp[i][2] 파

# 색깔은 행이 바뀔 때마다 겹치면 안됨 --> 즉 크로스 합
# https://www.youtube.com/watch?v=QZXb1JWQ2Gc

import sys

input = sys.stdin.readline
N = int(input())
arr = [list(map(int, input().split())) for _ in range(N)]
dp = [[0] * 3 for _ in range(N)]
dp[0] = arr[0]

for i in range(N):
    dp[i][0] = min(dp[i-1][1], dp[i-1][2]) + arr[i][0]
    dp[i][1] = min(dp[i-1][0], dp[i-1][2]) + arr[i][1]
    dp[i][2] = min(dp[i-1][0], dp[i-1][1]) + arr[i][2]

print(min(dp[N-1]))

File: 백준/Silver/1149. RGB거리/RGB거리.java
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.BufferedReader;


public class Main {	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int N = Integer.parseInt(br.readLine());
		int[][] arr = new int[N][N];
		for (int r = 0; r < N; r++) {
			String[] cols = br.readLine().split(" ");
			for (int c = 0; c < 3; c++) {
				arr[r][c] = Integer.parseInt(cols[c]);
			}
		}
		lets_figure_out_min_cost(N, arr);
	}
	
	
	
	static void lets_figure_out_min_cost(int N, int[][] arr) {
		int[][] DP = new int[N][3];
		DP[0][0] = arr[0][0];
		DP[0][1] = arr[0][1];
		DP[0][2] = arr[0][2];
		
		// 색깔이 겹치면 안된다 -> 각각 보자
		for (int i = 1; i < N; i++) {
			DP[i][0] = Math.min(DP[i-1][1], DP[i-1][2]) + arr[i][0];
			DP[i][1] = Math.min(DP[i-1][0], DP[i-1][2]) + arr[i][1];
			DP[i][2] = Math.min(DP[i-1][1] , DP[i-1][0]) + arr[i][2];
		}
		
		int res = 1001;
		res = Math.min(DP[N-1][0], Math.min(DP[N-1][1], DP[N-1][2]));
		System.out.println(res);
		
	}
	
	
}